### DISCUSSION TITLE
Stories not loading in one of multiple instances in a monorepo (with shared modules)

### DISCUSSION THREAD

**Discussion by beardage** [storybookjs/storybook#24687] [Help] [answered]
![image](https://github.com/storybookjs/storybook/assets/5056251/491799f1-8305-49f1-b25f-5c335c306657)
I'm currently working through the warnings, which seem to be [related to this issue](https://github.com/storybookjs/storybook/issues/21316), but no errors to be seen. Is there a good workflow for debugging rendering issues? 
All stories perpetually load in this instance since all npm modules were moved to the root level (unclear if this is related but seemingly.) Could also be related to [this issue](https://github.com/storybookjs/storybook/issues/21710)?

Edit: also confirmed that on the main branch of this repo (without the restructuring of modules in the monorepo), those warnings still appear but the stories load without issue.

**Comment by dannyhw** [comment]
I would assume its not related to that warning. Are there no errors on the storybook server? Also if you can produce a reproduction of the issue that would make it easier to debug.

**Reply by beardage** [reply]
Unfortunately there are no errors on the storybook server. I have been able to reduce the warnings without success at getting stories to render. I'll try to put some time into creating a reproduction. Are there any flags I can use to debug a rendering issue? I'm surprised there's no error reporting for this.

**Comment by woodsboe** [comment]
Could you provide some more technical specs about your solution?
- What kind of monorepo?
- Frontend framework?
- Storybook configs?
- etc.

**Reply by beardage** [reply]
1. NPM workspaces
2. Nextjs 12.2.2
4. 
`apps/web/.storybook/main.ts` (broken instance)
```
import type { StorybookConfig } from "@storybook/nextjs";

import { join, dirname } from "path";

const config: StorybookConfig = {
  stories: ["../src/**/*.mdx", "../src/**/*.stories.@(js|jsx|mjs|ts|tsx)"],
  addons: [
    "@chakra-ui/storybook-addon",
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
  ],
  framework: {
    name: "@storybook/nextjs",
    options: {},
  },
  docs: {
    autodocs: "tag",
  },
};
export default config;
```

`packages/ui/.storybook/main.ts` (working instance)
```
import type { StorybookConfig } from "@storybook/nextjs";

import { join, dirname } from "path";

const config: StorybookConfig = {
  stories: ["../src/**/*.mdx", "../src/**/*.stories.@(js|jsx|mjs|ts|tsx)"],
  addons: [
    "@chakra-ui/storybook-addon",
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
  ],
  framework: {
    name: "@storybook/nextjs",
    options: {},
  },
  docs: {
    autodocs: "tag",
  },
};
export default config;
```

root package.json includes:
```
...
"dependencies": {
    "next": "12.2.2",
    "webpack": "^5.86.0",
    "zod": "3.22.4
},
"devDependencies": {
   "@storybook/addon-a11y": "7.5.2",
    "@storybook/addon-essentials": "7.5.2",
    "@storybook/addon-interactions": "7.5.2",
    "@storybook/addon-links": "7.5.2",
    "@storybook/blocks": "7.5.2",
    "@storybook/jest": "^0.1.0",
    "@storybook/react": "7.5.2",
    "@storybook/nextjs": "7.5.2",
    "@storybook/test-runner": "^0.12.0",
    "@storybook/testing-library": "0.2.0",
    "storybook": "7.5.2"
    "msw": "^1.2.1",
    "msw-storybook-addon": "^1.8.0"
}
...
```

**Reply by woodsboe** [reply]
Hi @beardage,
Thanks for posting the configs.

I'm running a similar setup (Turborepo, NextJS and Storybook), but there's a slide difference in my config files.

You might be missing the `getAbsolutePath()` function, which you can see below.
The function is part of the standard Storybook installation if you install Storybook in a monorepo.

Try incorporating the `getAbsolutePath()` function into your configs; hopefully, it'll fix your problem.  

```typescript
import type { StorybookConfig } from "@storybook/nextjs";

import { join, dirname } from "path";

/**
 * This function is used to resolve the absolute path of a package.
 * It is needed in projects that use Yarn PnP or are set up within a monorepo.
 */
function getAbsolutePath(value: string): any {
  return dirname(require.resolve(join(value, "package.json")));
}
const config: StorybookConfig = {
  stories: [
    "../stories/**/*.mdx",
    "../stories/**/*.stories.@(js|jsx|mjs|ts|tsx)",
  ],
  addons: [
    getAbsolutePath("@storybook/addon-links"),
    getAbsolutePath("@storybook/addon-essentials"),
    getAbsolutePath("@storybook/addon-onboarding"),
    getAbsolutePath("@storybook/addon-interactions"),
  ],
  framework: {
    name: getAbsolutePath("@storybook/nextjs"),
    options: {},
  },
  docs: {
    autodocs: "tag",
  },
};
export default config;
```

**Reply by beardage** [reply, marked as answer]
Thank you for sharing your `main.ts` file! I actually solved it while testing your solution, but it wasn't actually the absolute path that was the issue. For future discussion-searchers:

If the only thing that doesn't load is the preview window, like my screenshot above, the problem may actually be in preview.js/ts.

The issue ended up being the implementation of loaders in preview.ts, specifically mswLoader (line 38 in screenshot below) from msw-storybook-addon. Msw was configured to use the service worker in the app's public folder. I incorrectly assumed it was sharing the public folder from the app (`apps/web/public`) which had a much newer version of the generated mockServiceWorker.js than the version within `apps/web/.storybook/public`. Once those were in sync, the stories rendered as expected.  

![image](https://github.com/storybookjs/storybook/assets/5056251/566d1a00-393a-4c14-accc-88c0dfefd225)

Thank you to everyone for the assistance. Maybe I should file an issue for the lack of reporting on this issue on the addon's repo?

**Reply by valentinpalkovic** [reply]
Yes, please feel free to do so!

**Comment by asanger** [comment]
For everyone in this thread, I'm running a similar setup with Nextjs, a web workspace, and then a shared workspace for common components. When I  am creating a story in `/web` that imports a component from `shared`, that component always comes back as `undefined`. If I import a component from `/web`, it works as expected. I was wondering if any of you experienced that with this sort of setup. I have a simple demo repo to reproduce it here: https://github.com/asanger/storybook-error-demo
